src/bindings/python/flux/cli/plugin.py

```diff
@@ -0,0 +1,249 @@
+    but this can be overridden by setting an environment variable,
+    ``FLUX_CLI_PLUGINPATH``.
+    """
+
+    def __init__(self):
+        self.plugins = []
+        self.plugin_options = {}
+        etc = conf_builtin_get("confdir")
+        if getenv("FLUX_CLI_PLUGINPATH"):
+            self.plugindir = getenv("FLUX_CLI_PLUGINPATH")
+        else:
+            if etc is None:
+                raise ValueError("failed to get builtin confdir")
+            self.plugindir = f"{etc}/cli/plugins"
+
+    def _add_plugins(self, module, program):
+        entries = []
+        for attr in dir(module):
+            entries.append(getattr(module, attr))
+        for entry in entries:
+            # only process entries that are an instance of type (i.e. a class)
+            # and are a subclass of CLIPlugin but not the base class:
+            if (
+                isinstance(entry, type)
+                and issubclass(entry, CLIPlugin)
+                and CLIPlugin != entry
+            ):
+                curr = entry(program)
+                self.plugins.append(curr)
+                for option, usage in curr.options.items():
+                    self._add_option(option, usage)
+
+    def _add_option(self, opt, usage):
+        self.plugin_options[opt] = usage
+
+    def load_plugins(self, program):
+        """Load all cli plugins from the standard path"""
+        for path in glob.glob(f"{self.plugindir}/*.py"):
+            self._add_plugins(import_path(path), program)
+        return self
+
+    def get_plugin_options(self):
+        """Return all plugin option keys from self.plugins"""
+        return self.plugin_options.keys()
+
+    def get_plugin_usages(self):
+        """Return a string that has all self.plugin usage messages"""
+        usages = None
+        if self.plugins:
+            usages = "Options provided by plugins:\n"
+        for opt, usage in self.plugin_options.items():
+            p_usage = textwrap.fill(usage, width=60).splitlines()
+            usages += f"  {opt:<20}{p_usage[0]}\n"
+            if len(p_usage) > 1:
+                for line in p_usage[1:]:
+                    usages += f"{' ' * 22}{line}\n"
+        return usages
+
+    def preinit(self, args):
+        """Call all plugin ``preinit`` callbacks.
+
+        Returns:
+            values (:py:obj:CLIPluginNamespace): A namespace of key-value
+                pairs for loaded plugin options.
+        """
+        values = CLIPluginNamespace()
+        for plugin in self.get_plugin_options():
+            setattr(values, str(plugin), None)
+        if args.plugin:
+            for provided_arg in args.plugin:
+                key, val = provided_arg.get_value()
+                if key in self.get_plugin_options():
+                    setattr(values, key, val)
+                else:
+                    raise ValueError(
+                        f"Unsupported option provided to -P/--plugin: {key}"
+                    )
```

Comment:
I think it might be more clear to split out the handling of `--plugin` arguments into a separate method, then store the `CLIPluginNamespace` as an attribute in the registry, instead of constructing it on demand here in preinit().

Then the plugin args namespace doesn't have to be exposed to the registry users, and e.g. when `modify_jobspec()` is called, instead of passing "`values`" (which is really the CLIPluginNamespace), that argument can be dropped and `self.args` or `self.plugin_args` can be passed to each `plugin.modify_jobspec()`.